src/main/java/uk/ac/ebi/ampt2d/accession/service/MonotonicAccessionGenerator.java

```diff
@@ -0,0 +1,144 @@
+        this.contiguousIdBlockRepository = contiguousIdBlockRepository;
+        this.blockState = new BlockState();
+        loadIncompleteBlocks();
+    }
+
+    private void loadIncompleteBlocks() {
+        List<ContiguousIdBlock> uncompletedBlocks = getUncompletedBlocksForThisInstanceOrdered();
+        //Insert as available ranges
+        for (ContiguousIdBlock block : uncompletedBlocks) {
+            blockState.addBlock(block);
+        }
+    }
+
+    /**
+     * TODO this can be changed to a query to recover and avoid the filter
+     */
+    private List<ContiguousIdBlock> getUncompletedBlocksForThisInstanceOrdered() {
+        try (Stream<ContiguousIdBlock> reservedBlocksOfThisInstance = contiguousIdBlockRepository
+                .findAllByCategoryIdAndApplicationInstanceIdOrderByEndAsc(categoryId, applicationInstanceId)) {
+            return reservedBlocksOfThisInstance.filter(ContiguousIdBlock::isNotFull).collect(Collectors.toList());
+        }
+    }
+
+    /**
+     * This function will recover the internal state of committed elements and will remove them from the available
+     * ranges.
+     *
+     * @param committedElements
+     * @throws AccessionIsNotPending
+     */
+    public synchronized void recoverState(long[] committedElements) throws AccessionIsNotPending {
+        blockState.recoverState(committedElements);
+    }
+
+    public synchronized long[] generateAccessions(int numAccessionsToGenerate) {
+        long[] accessions = new long[numAccessionsToGenerate];
+        reserveNewBlocksUntilSizeIs(numAccessionsToGenerate);
+
+        int i = 0;
+        while (i < numAccessionsToGenerate) {
+            int remainingAccessionsToGenerate = numAccessionsToGenerate - i;
+            long[] ids = blockState.pollNextMonotonicValues(remainingAccessionsToGenerate);
+            System.arraycopy(ids, 0, accessions, i, ids.length);
+            i += ids.length;
+        }
+        assert (i == numAccessionsToGenerate);
+
+        return accessions;
+    }
+
+    /**
+     * Ensures that the available ranges queue hold @param totalAccessionsToGenerate or more elements
+     *
+     * @param totalAccessionsToGenerate
+     */
+    private synchronized void reserveNewBlocksUntilSizeIs(int totalAccessionsToGenerate) {
+        while (blockState.isAvailableSpaceLessThan(totalAccessionsToGenerate)) {
+            try {
+                ExponentialBackOff.execute(() -> reserveNewBlock(categoryId, applicationInstanceId, blockSize), 10, 30);
+            } catch (RuntimeException e) {
+                // Ignore, max backoff have been reached, we will try again until we can reserve blocks
+            }
+        }
+    }
+
+    @Transactional(isolation = Isolation.SERIALIZABLE)
+    protected synchronized void reserveNewBlock(String categoryId, String instanceId, long size) {
+        ContiguousIdBlock lastBlock = contiguousIdBlockRepository.findFirstByCategoryIdOrderByEndDesc(categoryId);
+        if (lastBlock != null) {
```

Comment:
what if the last block returned by the repository is already full?

Reply:
We will have a useless block in memory that will never get removed as the commit operation :/ good catch, needs a test